### PR TITLE
Unconditionally refresh storage options from config

### LINKED ISSUE #2217: State database doesn't save storage options
/kind bug

**Description**

State DB (Bolt) doesn't save storage options, therefore they are honored on the first run, but not on following runs.

**Steps to reproduce the issue:**

1. Configure a storage option (e.g. `ostree_repo = "/var/lib/ostree"`).

2.  Pull an image with debug level "looking" for the options output:  `sudo podman --log-level debug pull alpine:3.8 2>&1 | grep "overlay: "`

3.  Pull another image with debug level "looking" for the options output:  `sudo podman --log-level debug pull alpine:3.7 2>&1 | grep "overlay: "`

**Describe the results you received:**

On first pull:
```
time="..." level=debug msg="overlay: ostree_repo=/var/lib/ostree" 
time="..." level=debug msg="overlay: override_kernelcheck=true" 
time="..." level=warning msg="Not using native diff for overlay, this may cause degraded performance for building images: failed to mount overlay: invalid argument" 
```

References available in OStree repository (`ostree refs --repo=/var/lib/ostree`):
```
containers-storage/7bff100f35cb359a368537bb07829b055fe8e0b1cb01085a3a628ae9c187c7b8
```

On second pull:
```
<nothing at all>
```

References available in OStree repository (`ostree refs --repo=/var/lib/ostree`):
```
containers-storage/7bff100f35cb359a368537bb07829b055fe8e0b1cb01085a3a628ae9c187c7b8
```

**Describe the results you expected:**

On first pull:
```
time="..." level=debug msg="overlay: ostree_repo=/var/lib/ostree" 
time="..." level=debug msg="overlay: override_kernelcheck=true" 
time="..." level=warning msg="Not using native diff for overlay, this may cause degraded performance for building images: failed to mount overlay: invalid argument" 
```

References available in OStree repository (`ostree refs --repo=/var/lib/ostree`):
```
containers-storage/7bff100f35cb359a368537bb07829b055fe8e0b1cb01085a3a628ae9c187c7b8
```

On second pull:
```
time="..." level=debug msg="overlay: ostree_repo=/var/lib/ostree" 
time="..." level=debug msg="overlay: override_kernelcheck=true" 
time="..." level=warning msg="Not using native diff for overlay, this may cause degraded performance for building images: failed to mount overlay: invalid argument" 
```

References available in OStree repository (`ostree refs --repo=/var/lib/ostree`):
```
containers-storage/d6da3c54c8f3a43215d51cc0d43eff9759624996e40aaf11e4be5a555c038c32
containers-storage/7bff100f35cb359a368537bb07829b055fe8e0b1cb01085a3a628ae9c187c7b8
```

**Additional information you deem important (e.g. issue happens only occasionally):**

If the state DB (`bolt_state.db`) is removed after the first run, then the output is the expected one.

**Output of `podman version`:**

```
Version:       1.0.0
Go Version:    go1.11.4
Git Commit:    ""
Built:         Mon Jan 14 23:12:09 2019
OS/Arch:       linux/amd64
```

**Output of `podman info`:**

```
host:
  BuildahVersion: 1.6-dev
  Conmon:
    package: Unknown
    path: /usr/libexec/crio/conmon
    version: 'conmon version 1.13.0, commit: e8a2525c2e7f5ab057d5a2b5f1950be5643d8053-dirty'
  Distribution:
    distribution: arch
    version: unknown
  MemFree: 245493760
  MemTotal: 3923992576
  OCIRuntime:
    package: Unknown
    path: /usr/bin/runc
    version: |-
      runc version 1.0.0-rc5+dev
      commit: 079817cc26ec5292ac375bb9f47f373d33574949
      spec: 1.0.1-dev
  SwapFree: 8188518400
  SwapTotal: 8204779520
  arch: amd64
  cpus: 4
  hostname: ...
  kernel: 4.20.3-arch1-1-ARCH
  os: linux
  rootless: false
  uptime: 2h 24m 40.3s (Approximately 0.08 days)
insecure registries:
  registries: []
registries:
  registries:
  - docker.io
  - registry.fedoraproject.org
  - quay.io
  - registry.access.redhat.com
  - registry.centos.org
store:
  ConfigFile: /etc/containers/storage.conf
  ContainerStore:
    number: 0
  GraphDriverName: overlay
  GraphOptions: null
  GraphRoot: /var/lib/containers/storage
  GraphStatus:
    Backing Filesystem: extfs
    Native Overlay Diff: "false"
    Supports d_type: "true"
  ImageStore:
    number: 2
  RunRoot: /var/run/containers/storage
```

Whereas the value of `GraphOptions` after first run is
```
  - overlay.ostree_repo=/var/lib/ostree
  - overlay.mountopt=nodev
  - overlay.override_kernel_check=true
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Bare metal installation of Arch Linux. Libpod from AUR package: https://aur.archlinux.org/packages/libpod/

### PR DESCRIPTION
**Pull request by mheon**
Due to our unconditionally setting some storage options, we are not always reading storage options from storage.conf. This can lead to some fields in the storage config (most notably extra storage options) being ignored, despite being set in storage.conf.

Resolve this by unconditionally refreshing our storage config from storage.conf (this was previously only done for rootless Podman)

Fixes #2217

### CODE REVIEW

**Comment by mheon** [comment]
Hm. Is it auto-approving PRs from approvers?

**Comment by rhatdan** [comment]
LGTM
@giuseppe PTAL

**Comment by Silvanoc** [comment]
LGTM

Branch tested on Arch Linux and behaves as expected.

**Comment by giuseppe** [comment]
it seems to break the creation of `~/.config/containers/storage.conf` when it doesn't exist.

Reproducer:

```shell
$ rm ~/.config/containers/storage.conf
$ bin/podman run --rm -ti alpine echo hi
$ cat ~/.config/containers/storage.conf
cat: /home/gscrivano/.config/containers/storage.conf: No such file or directory
```

File: pkg/util/utils.go [hunk 1]
```diff
@@ -301,36 +301,36 @@ func getTomlStorage(storeOptions *storage.StoreOptions) *tomlConfig {
...
+	}
 
-		storageConf := StorageConfigFile()
-		if _, err := os.Stat(storageConf); err == nil {
-			defaultRootlessRunRoot := storageOpts.RunRoot
-			defaultRootlessGraphRoot := storageOpts.GraphRoot
-			storageOpts = storage.StoreOptions{}
-			storage.ReloadConfigurationFile(storageConf, &storageOpts)
+	storageConf := StorageConfigFile()
+	if _, err := os.Stat(storageConf); err == nil {
```
Comment (giuseppe) [inline pkg/util/utils.go]: the issue is that `err` exists only in the `if` scope.  if we change it to "_, err = ...` then it works fine

Comment (mheon) [reply pkg/util/utils.go]: Fixed

**Comment by giuseppe** [comment]
LGTM

**Comment by Silvanoc** [comment]
LGTM

Current version of the branch behaves as expected on my tests for both root and rootless containers.

**Pull request merged**